### PR TITLE
Wormhole message reuse via post_message_unreliable

### PR DESCRIPTION
**Pull request by drozdziak1**
this changeset switches us to a PDA scheme for wormhole messages, accounts for reusing those accounts and conditionally switches between post_message/post_message_unreliable depending on whether a given message account was already initialized.

### CODE REVIEW

File: solana/pyth2wormhole/client/src/message.rs
Comment (jayantk) [inline solana/pyth2wormhole/client/src/message.rs:25]: the name of this struct is confusing because `index` could mean a number, and each account has a numeric id. i suggest `P2WMessageAccountQueue` ?

**Review by jayantk — APPROVED** [review]
this looks great to me. I'd like to get another pair of eyes on it before merging for security's sake, but as far as I can tell looks really solid.

Agree with your slack comment about making the reuse duration configurable, but that can be done in a separate PR.

File: solana/pyth2wormhole/client/tests/test_migrate.rs
Comment (jayantk) [inline solana/pyth2wormhole/client/tests/test_migrate.rs:186]: i take it we don't have clippy / pre-commit here which is why there's all this formatting noise?

we should install clippy. it's annoying to review PRs with these format changes mixed in.

File: solana/pyth2wormhole/program/src/attest.rs
Comment (ali-bahjati) [inline solana/pyth2wormhole/program/src/attest.rs]: does realloc make sure that the account will remain rent exempt? 

I'm thinking that the size might be different if we have batches of different size and it can cause doing resizes more than necessary. 


minor: do we need to handle resizes? Can't we use a version field in the instruction input and `P2WMessageDrvData`? Or maybe a "{size}" as a seed?

File: solana/pyth2wormhole/program/src/attest.rs
Comment (ali-bahjati) [reply solana/pyth2wormhole/program/src/attest.rs:283]: I don't know whether it was for local debug or not. But these `trace!` macros won't do anything on releases (it's only for debug).

File: solana/pyth2wormhole/program/src/attest.rs
Comment (ali-bahjati) [inline solana/pyth2wormhole/program/src/attest.rs:271]: I couldn't understand that why we are using reliable message post when account is not initialized. I think unreliable_post_message also creates an account if needed:
https://github.com/wormhole-foundation/wormhole/blob/dev.v2/solana/bridge/program/src/api/post_message.rs#L160

**Review by ali-bahjati — APPROVED** [review]
Very nice work 🔥 

I have one concern about doing more resizes than necessary that I asked, apart from that lgtm.

File: solana/pyth2wormhole/program/src/attest.rs
Comment (tompntn) [inline solana/pyth2wormhole/program/src/attest.rs]: Am I missing something (likely!) or is this equivalent to just doing `accs.wh_message.0 .0 .0.realloc(payload.len(), false)`?

File: solana/pyth2wormhole/program/src/attest.rs
Comment (tompntn) [inline solana/pyth2wormhole/program/src/attest.rs]: Curious - why do we only resize if the account is un-initialized? Does the account ever get reset back to un-initialized, or do we only ever resize once?

File: solana/pyth2wormhole/program/src/attest.rs
Comment (Reisen) [inline solana/pyth2wormhole/program/src/attest.rs:278]:
```suggestion
            accs.wh_message.info().realloc(
```

File: solana/pyth2wormhole/client/src/message.rs
Comment (tompntn) [inline solana/pyth2wormhole/client/src/message.rs]: Would it be a good idea to set an upper bound on the number of accounts we can create, as a safety net in case a bug causes a huge number of accounts to be created?

File: solana/pyth2wormhole/client/src/message.rs
Comment (tompntn) [inline solana/pyth2wormhole/client/src/message.rs]: Should we check here that the ID hasn't overflowed? Unlikely but paranoid 😉

File: solana/pyth2wormhole/program/src/attest.rs
Comment (Reisen) [inline solana/pyth2wormhole/program/src/attest.rs:297]: Is there a reason `bridge::instructions::post_message_unreliable` can’t be used over manual construction? Same question regarding `post_message` below (though also curious about Ali’s question above, could this just be a call to unreliable?)

File: solana/pyth2wormhole/program/src/attest.rs
Comment (Reisen) [inline solana/pyth2wormhole/program/src/attest.rs:302]:
```suggestion
                AccountMeta::new(*accs.wh_message.info().key, true),
```

File: solana/pyth2wormhole/program/src/attest.rs
Comment (Reisen) [inline solana/pyth2wormhole/program/src/attest.rs:328]:
```suggestion
                AccountMeta::new(*accs.wh_message.info().key, true),
```

**Review by Reisen — CHANGES_REQUESTED** [review]

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs:283]: Yeah, The trace!-s evaluate to nothing without the `trace` feature and can speed up debugging

File: solana/pyth2wormhole/client/tests/test_migrate.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/client/tests/test_migrate.rs:186]: I will try to undo the formatting-related changes. Sorry about that!

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs]: I don't think it takes care of rent exemptions. nice catch! I think wormhole would start complaining the next time we expand the format/symbol count per batch, which could become a nasty debug session. Thanks!

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs:271]: Nice, this simplifies the critical path. Thanks!

File: solana/pyth2wormhole/client/src/message.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/client/src/message.rs]: I think we'll go out of memory many orders of magnitude before this counter gets wrecked

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs:297]: It could be just unreliable and I probably could use the function. Thanks!

File: solana/pyth2wormhole/client/src/message.rs
Comment (tompntn) [reply solana/pyth2wormhole/client/src/message.rs]: I'm more concerned with the downstream effects on the Wormhole guardians running out of space than any sort of memory issues with the attester.

File: solana/pyth2wormhole/client/src/message.rs
Comment (ali-bahjati) [inline solana/pyth2wormhole/client/src/message.rs:80]: very nice ser 🙇

File: solana/pyth2wormhole/client/src/message.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/client/src/message.rs]: Done

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs]: Added a routine that makes sure exemption is maintained.

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs]: Done, this looks so much nicer now!

File: solana/pyth2wormhole/client/src/message.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/client/src/message.rs]: Your comment nerdsniped me real good. Ultimately I don't want people waking up on-call over a bullshit human error either. I added a pretty good test and will add the limit.

Pragmatism: 1
Me: 0

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs]: Done!

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs]: removed in favor of the upstream instruction generator

File: solana/pyth2wormhole/program/src/attest.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/program/src/attest.rs]: Done!

File: solana/pyth2wormhole/client/src/message.rs
Comment (drozdziak1) [reply solana/pyth2wormhole/client/src/message.rs:88]: Thanks!

File: solana/pyth2wormhole/client/src/message.rs
Comment (tompntn) [reply solana/pyth2wormhole/client/src/message.rs]: Thank-you!

**Review by Reisen — APPROVED** [review]

**Pull request merged**
